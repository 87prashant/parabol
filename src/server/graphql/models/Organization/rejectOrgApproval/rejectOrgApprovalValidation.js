import legitify from 'universal/validation/legitify';
import {orgRejectionReason, requiredId} from 'universal/validation/templates';

export default function rejectOrgApprovalValidation() {
  return legitify({
    dbNotificationId: requiredId,
    reason: orgRejectionReason
  });
}
