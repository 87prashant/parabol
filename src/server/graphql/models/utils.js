import {GraphQLNonNull, GraphQLInputObjectType} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import mime from 'mime-types';
import {
  APP_CDN_USER_ASSET_SUBDIR,
  APP_MAX_AVATAR_FILE_SIZE
} from 'universal/utils/constants';
import protocolRelativeUrl from 'server/utils/protocolRelativeUrl';
import {s3SignPutObject} from 'server/utils/s3';
import path from 'path';

// Stringify an object to handle multiple errors
// Wrap it in a new Error type to avoid sending it twice via the originalError field
export const errorObj = obj => new Error(JSON.stringify(obj));

export const handleSchemaErrors = (errors) => {
  if (Object.keys(errors).length > 0) {
    throw errorObj(errors);
  }
};

// VERY important, otherwise eg a user could "create" a new team with an existing teamId & force join that team
export const ensureUniqueId = async (table, id) => {
  const r = getRethink();
  const res = await r.table(table).get(id);
  if (res) {
    throw errorObj({type: 'unique id collision'});
  }
};

export const ensureUserInOrg = async (userId, orgId) => {
  const r = getRethink();
  const inOrg = await r.table('User').get(userId)('orgs').contains(orgId);
  if (!inOrg) {
    throw errorObj({type: `user ${userId} does not belong to org ${orgId}`});
  }
  return true;
};

// if the add & update schemas have different required fields, use this
export const nonnullifyInputThunk = (name, inputThunk, requiredFieldNames) => {
  return new GraphQLInputObjectType({
    name,
    fields: () => {
      const newFields = inputThunk();
      requiredFieldNames.forEach(fieldName => {
        newFields[fieldName].type = new GraphQLNonNull(newFields[fieldName].type);
      });
      return newFields;
    }
  });
};

function getFields(context, astsParams = context.fieldNodes) {
  // for recursion...Fragments doesn't have many sets...
  const asts = Array.isArray(astsParams) ? astsParams : [astsParams];

  // get all selectionSets
  const selectionSets = asts.reduce((selections, source) => {
    selections.push(...source.selectionSet.selections);
    return selections;
  }, []);

  // return fields
  return selectionSets.reduce((list, ast) => {
    switch (ast.kind) {
      case 'Field' :
        // eslint-disable-next-line no-param-reassign
        list[ast.name.value] = true;
        return list;
      case 'InlineFragment':
        return {
          ...list,
          ...getFields(context, ast)
        };
      case 'FragmentSpread':
        return {
          ...list,
          ...getFields(context, context.fragments[ast.name.value])
        };
      default:
        throw new Error('Unsuported query selection');
    }
  }, {});
}

export function getRequestedFields(refs) {
  const fieldsObj = getFields(refs);
  return Object.keys(fieldsObj);
}

export function firstChange(possiblyUpdatedResult) {
  if (possiblyUpdatedResult.changes.length) {
    if (possiblyUpdatedResult.changes.length > 1) {
      console.warn('firstChange() detects more than 1 change, returning 1st.');
    }
    return possiblyUpdatedResult.changes[0];
  }
  return { new_val: undefined, old_val: undefined };
}

export function updatedOrOriginal(possiblyUpdatedResult, original) {
  /*
   * There will only be changes to return if there were changes made to the
   * DB. Therefore, we've got to check.
   */
  return firstChange(possiblyUpdatedResult).new_val || original;
}

export function getOldVal(possiblyUpdatedResult) {
  return firstChange(possiblyUpdatedResult).old_val;
}

export function getNewVal(resultWithReturnChanges) {
  return firstChange(resultWithReturnChanges).new_val;
}

export function makeEnumValues(constArr) {
  return constArr.reduce((values, constant) => {
    values[constant] = {value: constant};
    return values;
  }, {})
}

export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw errorObj({_error: 'CDN_BASE_URL environment variable is not defined'});
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw errorObj({_error: 'file must be an image'});
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw errorObj({_error: `unable to determine extension for ${contentType}`});
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw errorObj({_error: 'avatar image is too large'});
  }
  return ext;
}

export function getS3PutUrl(contentType, contentLength, partialPath) {
  const parsedUrl = protocolRelativeUrl.parse(process.env.CDN_BASE_URL);
  const pathname = path.join(parsedUrl.pathname,
    APP_CDN_USER_ASSET_SUBDIR,
    partialPath
  );
  return s3SignPutObject(
    pathname,
    contentType,
    contentLength,
    'public-read'
  );
}
