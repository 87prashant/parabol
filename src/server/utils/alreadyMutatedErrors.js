import sendAuthRaven from 'server/utils/sendAuthRaven';

export const sendAlreadyJoinedIntegrationError = (authToken, integrationId) => {
  const breadcrumb = {
    message: 'You are already a part of this integration',
    category: 'Already mutated',
    data: {integrationId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
};

export const sendAlreadyUpdatedIntegrationError = (authToken, integrationId) => {
  const breadcrumb = {
    message: 'Integration was already updated',
    category: 'Already mutated',
    data: {integrationId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
};

export const sendAlreadyArchivedTeamError = (authToken, teamId) => {
  const breadcrumb = {
    message: 'Team was already archived',
    category: 'Team Already Archived',
    data: {teamId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
}

export const sendAlreadyInactivatedUserError = (authToken, userId) => {
  const breadcrumb = {
    message: 'That user is already inactive. cannot inactivate twice',
    category: 'Already inactive',
    data: {userId}
  };
  return sendAuthRaven(authToken, 'Easy there', breadcrumb);
}
