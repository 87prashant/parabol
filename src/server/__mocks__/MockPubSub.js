import * as getPubSub from 'server/utils/getPubSub';

const fieldsToSerialize = {
  invitation: [
    'channelId',
    'message.data.invitation.id',
    'message.data.invitation.email',
    'message.data.invitation.hashedToken',
    'message.data.invitation.id',
    'message.data.invitation.invitedBy',
    'message.data.invitation.teamId'
  ],
  invitationAdded: [
    'channelId',
    'message.invitationAdded.invitation.id',
    'message.invitationAdded.invitation.email',
    'message.invitationAdded.invitation.hashedToken',
    'message.invitationAdded.invitation.id',
    'message.invitationAdded.invitation.invitedBy',
    'message.invitationAdded.invitation.teamId'
  ],
  invitationRemoved: [
    'channelId',
    'message.invitationRemoved.invitation.id',
    'message.invitationRemoved.invitation.email',
    'message.invitationRemoved.invitation.hashedToken',
    'message.invitationRemoved.invitation.inviteToken',
    'message.invitationRemoved.invitation.invitedBy',
    'message.invitationRemoved.invitation.teamId'
  ],
  newAuthToken: [
    'channelId',
    'message.newAuthToken'
  ],
  notification: [
    'channelId',
    'message.data.notifications.id',
    'message.data.notifications.orgId',
    'message.data.notifications.teamId',
    'message.data.notifications.userIds',
    'message.data.notifications.inviteeEmail'
  ],
  notificationsCleared: [
    'channelId',
    'message.notificationsCleared.deletedIds'
  ],
  orgApprovalRemoved: [
    'channelId',
    'message.orgApprovalRemoved.orgApproval.approvedBy',
    'message.orgApprovalRemoved.orgApproval.deniedBy',
    'message.orgApprovalRemoved.orgApproval.email',
    'message.orgApprovalRemoved.orgApproval.id',
    'message.orgApprovalRemoved.orgApproval.orgId',
    'message.orgApprovalRemoved.orgApproval.teamId'
  ],
  teamAdded: [
    'channelId',
    'message.teamAdded.notification.teamId',
    'message.teamAdded.teamId',
    'message.teamAdded.teamMemberId'
  ],
  teamUpdated: [
    'channelId',
    'message.teamUpdated.team.id',
    'message.teamUpdated.team.orgId'
  ],
  teamMemberAdded: [
    'channelId',
    'message.teamMemberAdded.teamMember.id',
    'message.teamMemberAdded.teamMember.email',
    'message.teamMemberAdded.teamMember.teamId',
    'message.teamMemberAdded.teamMember.userId'
  ]
};

const fieldsToSort = {
  notificationsAdded: [
    'message.notificationsAdded.notifications.0.teamName',
    'message.notificationsAdded.notifications.0.type'
  ],
  orgApprovalRemoved: [
    'message.orgApprovalRemoved.orgApproval.createdAt'
  ]
};

const unsortables = new Set(['invitationAdded']);


const getPath = (str, obj) => str.split('.').reduce((o, i) => o[i], obj);

export default class MockPubSub {
  constructor() {
    getPubSub.default = () => this;
    this.db = {};
  }

  __serialize(dynamicSerializer) {
    const channels = Object.keys(this.db).sort();
    const snapshot = {};
    for (let i = 0; i < channels.length; i++) {
      const channel = channels[i];
      const doc = this.db[channel];
      const channelFields = fieldsToSerialize[channel];
      if (!channelFields) {
        throw new Error(`BAD MOCK: No fieldsToSerialize for pubsub channel ${channel}`);
      }
      snapshot[channel] = dynamicSerializer.toStatic(doc, channelFields, {constant: unsortables.has(channel)});
      // we don't care about the order, so make it repeatable
      const customSorts = fieldsToSort[channel] || [];
      const sortFields = customSorts.concat('channelId');
      snapshot[channel].sort((a, b) => {
        for (let ii = 0; ii < sortFields.length; ii++) {
          const sortField = sortFields[ii];
          const aVal = getPath(sortField, a);
          const bVal = getPath(sortField, b);
          if (aVal === bVal) continue;
          return aVal > bVal;
        }
        return 0;
      });
    }
    return snapshot;
  }

  publish(channel, message) {
    const [channelName, channelId] = channel.split('.');
    this.db[channelName] = this.db[channelName] || [];
    this.db[channelName].push({channelId, message});
  }
}

