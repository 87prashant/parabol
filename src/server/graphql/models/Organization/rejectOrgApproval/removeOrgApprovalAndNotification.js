import getRethink from 'server/database/rethinkDriver';
import {REQUEST_NEW_USER} from 'universal/utils/constants';

const removeOrgApprovalAndNotification = async (orgId, maybeEmails) => {
  const emails = Array.isArray(maybeEmails) ? maybeEmails : [maybeEmails];
  const r = getRethink();
  const {removedNotification} = await r({
    removedApproval: r.table('OrgApproval')
      .getAll(r.args(emails), {index: 'email'})
      .filter({orgId})
      .delete(),
    removedNotification: r.table('Notification')
      .getAll(orgId, {index: 'orgId'})
      .filter({
        type: REQUEST_NEW_USER
      })
      .filter((notification) => {
        return r.expr(emails).contains(notification('inviteeEmail'));
      })
      // get the inviterName
      .delete({returnChanges: true})('changes')
      .map((change) => change('old_val'))
      .default([])
  })
  return removedNotification;
};

export default removeOrgApprovalAndNotification;
