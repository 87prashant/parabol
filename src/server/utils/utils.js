import mime from 'mime-types';
import {APP_MAX_AVATAR_FILE_SIZE} from '../../universal/utils/constants';

// todo put this in a legitify schema
export function validateAvatarUpload(contentType, contentLength) {
  if (typeof process.env.CDN_BASE_URL === 'undefined') {
    throw new Error('CDN_BASE_URL environment variable is not defined');
  }
  if (!contentType || !contentType.startsWith('image/')) {
    throw new Error('file must be an image');
  }
  const ext = mime.extension(contentType);
  if (!ext) {
    throw new Error(`unable to determine extension for ${contentType}`);
  }
  if (contentLength > APP_MAX_AVATAR_FILE_SIZE) {
    throw new Error('avatar image is too large');
  }
  return ext;
}
