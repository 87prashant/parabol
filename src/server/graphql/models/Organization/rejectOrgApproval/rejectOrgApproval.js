import getRethink from 'server/database/rethinkDriver';
import {
  GraphQLNonNull,
  GraphQLBoolean,
  GraphQLString,
  GraphQLID
} from 'graphql';
import {getUserId, getUserOrgDoc, requireWebsocket, requireOrgLeader} from 'server/utils/authorization';
import {errorObj, handleSchemaErrors} from 'server/utils/utils';
import rejectOrgApprovalValidation from 'server/graphql/models/Organization/rejectOrgApproval/rejectOrgApprovalValidation';
import removeOrgApprovalAndNotification from 'server/graphql/models/Organization/rejectOrgApproval/removeOrgApprovalAndNotification';
import shortid from 'shortid';
import {DENY_NEW_USER, NOTIFICATION_ADDED} from 'universal/utils/constants';
import getPubSub from 'server/utils/getPubSub';

export default {
  type: GraphQLBoolean,
  description: 'Create a new team and add the first team member',
  args: {
    notificationId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The notification to which the Billing Leader is responding'
    },
    reason: {
      type: GraphQLString
    }
  },
  async resolve(source, args, {authToken, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    const userId = getUserId(authToken);
    const {notificationId} = args;
    requireWebsocket(socket);
    const notification = await r.table('Notification').get(notificationId)
      .pluck('orgId', 'varList')
      .default(null);
    if (!notification) {
      throw errorObj({reason: `Notification ${notificationId} no longer exists!`});
    }
    const {orgId, inviterUserId, inviteeEmail} = notification;
    const userOrgDoc = await getUserOrgDoc(userId, orgId);
    requireOrgLeader(userOrgDoc);

    // VALIDATION
    const {data: {reason}, errors} = rejectOrgApprovalValidation()(args);
    handleSchemaErrors(errors);

    // RESOLUTION
    const deniedByName = await r.table('User').get(userId)('preferredName').default('A Billing Leader');
    const notificationAdded = {
      id: shortid.generate(),
      type: DENY_NEW_USER,
      startAt: now,
      orgId,
      userIds: [inviterUserId],
      reason,
      deniedByName,
      inviteeEmail
    };

    await Promise.all([
      removeOrgApprovalAndNotification(orgId, inviteeEmail),
      r.table('Notification').insert(notificationAdded)
    ]);
    getPubSub().publish(`${NOTIFICATION_ADDED}.${inviterUserId}`, {notificationAdded});
    return true;
  }
};
