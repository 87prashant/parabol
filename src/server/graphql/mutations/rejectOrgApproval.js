import {GraphQLID, GraphQLNonNull, GraphQLString} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import rejectOrgApprovalValidation from 'server/graphql/mutations/helpers/rejectOrgApprovalValidation';
import RejectOrgApprovalPayload from 'server/graphql/types/RejectOrgApprovalPayload';
import removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import {getUserId, getUserOrgDoc, requireOrgLeader} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import {errorObj, handleSchemaErrors} from 'server/utils/utils';
import shortid from 'shortid';
import {DENY_NEW_USER, NOTIFICATION, ORG_APPROVAL, REMOVED} from 'universal/utils/constants';
import promiseAllObj from 'universal/utils/promiseAllObj';

export default {
  type: RejectOrgApprovalPayload,
  description: 'Reject an invitee from joining any team under your organization',
  args: {
    notificationId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The notification to which the Billing Leader is responding'
    },
    reason: {
      type: GraphQLString
    }
  },
  async resolve(source, args, {authToken, dataLoader, socketId: mutatorId}) {
    const r = getRethink();
    const now = new Date();
    const operationId = dataLoader.share();
    const subOptions = {operationId, mutatorId};

    // AUTH
    const {notificationId} = args;
    const viewerId = getUserId(authToken);
    const rejectionNotification = await r.table('Notification').get(notificationId)
      .pluck('orgId', 'inviterUserId', 'inviteeEmail')
      .default(null);
    if (!rejectionNotification) {
      throw errorObj({reason: `Notification ${notificationId} no longer exists!`});
    }
    const {orgId, inviterUserId, inviteeEmail} = rejectionNotification;
    const userOrgDoc = await getUserOrgDoc(viewerId, orgId);
    requireOrgLeader(userOrgDoc);

    // VALIDATION
    const {data: {reason}, errors} = rejectOrgApprovalValidation()(args);
    handleSchemaErrors(errors);

    // RESOLUTION
    const deniedByName = await r.table('User').get(viewerId)('preferredName').default('A Billing Leader');
    const denialNotification = {
      id: shortid.generate(),
      type: DENY_NEW_USER,
      startAt: now,
      orgId,
      userIds: [inviterUserId],
      reason,
      deniedByName,
      inviteeEmail
    };

    // TODO include mutatorId in publishes once we're completely on Relay
    const {removedApprovalsAndNotifications} = await promiseAllObj({
      removedApprovalsAndNotifications: removeOrgApprovalAndNotification(orgId, inviteeEmail, {deniedBy: viewerId}),
      addNotification: r.table('Notification').insert(denialNotification)
    });

    const {removedOrgApprovals, removedRequestNotifications} = removedApprovalsAndNotifications;
    const removedOrgApprovalIds = removedOrgApprovals.map(({id}) => id);

    // publish the removed org approval to the team
    removedOrgApprovals.forEach((orgApproval) => {
      const {id: orgApprovalId, teamId} = orgApproval;
      publish(ORG_APPROVAL, teamId, REMOVED, {orgApprovalId}, subOptions);
    });

    // publish the removed request notifications to all org leaders
    removedRequestNotifications.forEach((notification) => {
      const {userIds} = notification;
      userIds.forEach((userId) => {
        publish(NOTIFICATION, userId, REMOVED, {notification}, subOptions);
      });
    });
    return {removedOrgApprovalIds, removedRequestNotifications};
  }
};
