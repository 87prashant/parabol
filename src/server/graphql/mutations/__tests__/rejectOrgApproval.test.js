import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockPubSub from 'server/__mocks__/MockPubSub';
import makeDataLoader from 'server/__tests__/setup/makeDataLoader';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import getRethink from 'server/database/rethinkDriver';
import rejectOrgApproval from 'server/graphql/mutations/rejectOrgApproval';
import {BILLING_LEADER, REQUEST_NEW_USER} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

describe('rejectOrgApproval', () => {
  test('rejects the invitee from all teams that wanted them', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockPubSub = new MockPubSub();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB
      .init()
      .user(1, {
        userOrgs: [{role: BILLING_LEADER, id: mockDB.context.organization.id}]
      })
      .newOrgApproval({email: invitee.email})
      .newNotification(undefined, {
        type: REQUEST_NEW_USER,
        email: invitee.email
      });
    const notificationId = mockDB.context.notification.id;
    const reason = 'bad hombre';
    const billingLeader = mockDB.db.user[0];
    const authToken = mockAuthToken(billingLeader);
    const dataLoader = makeDataLoader(authToken);
    const teamId = mockDB.context.team.id;
    // TEST
    await rejectOrgApproval.resolve(undefined, {notificationId, reason}, {authToken, dataLoader});

    // VERIFY
    const db = await fetchAndSerialize(
      {
        orgApproval: r
          .table('OrgApproval')
          .getAll(teamId, {index: 'teamId'})
          .orderBy('email'),
        notification: r
          .table('Notification')
          .getAll(billingLeader.id, {index: 'userIds'})
          .orderBy('userIds')
      },
      dynamicSerializer
    );

    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
  });
});
