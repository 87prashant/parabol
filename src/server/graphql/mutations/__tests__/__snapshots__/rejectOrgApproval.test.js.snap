// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejectOrgApproval rejects the invitee from all teams that wanted them 1`] = `
Object {
  "notification": Array [],
  "orgApproval": Array [
    Object {
      "createdAt": 2017-01-01T07:00:00.000Z,
      "deniedBy": 1,
      "email": 2,
      "id": 0,
      "isActive": true,
      "orgId": 3,
      "status": "DENIED",
      "teamId": 4,
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`rejectOrgApproval rejects the invitee from all teams that wanted them 2`] = `
Object {
  "notification": Array [
    Object {
      "channelId": 1,
      "message": Object {
        "data": Object {
          "deniedNotificationIds": Array [
            11,
          ],
          "removedOrgApprovalIds": Array [
            12,
          ],
          "removedRequestNotifications": "UNTESTED",
          "type": "RejectOrgApprovalPayload",
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 5,
      "message": Object {
        "data": Object {
          "deniedNotificationIds": Array [
            11,
          ],
          "removedOrgApprovalIds": Array [
            12,
          ],
          "removedRequestNotifications": "UNTESTED",
          "type": "RejectOrgApprovalPayload",
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
    Object {
      "channelId": 5,
      "message": Object {
        "data": Object {
          "deniedNotificationIds": Array [
            11,
          ],
          "removedOrgApprovalIds": Array [
            12,
          ],
          "removedRequestNotifications": "UNTESTED",
          "type": "RejectOrgApprovalPayload",
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
  ],
  "orgApproval": Array [
    Object {
      "channelId": 4,
      "message": Object {
        "data": Object {
          "deniedNotificationIds": Array [
            11,
          ],
          "removedOrgApprovalIds": Array [
            12,
          ],
          "removedRequestNotifications": "UNTESTED",
          "teamId": 4,
          "type": "RejectOrgApprovalPayload",
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
  ],
}
`;
